java/adapter/avro/src/main/java/org/apache/arrow/adapter/avro/consumers/AvroStructConsumer.java

```diff
@@ -68,8 +71,11 @@ public boolean resetValueVector(StructVector vector) {
 
   void ensureInnerVectorCapacity(long targetCapacity) {
     for (FieldVector v : vector.getChildrenFromFields()) {
-      while (v.getValueCapacity() < targetCapacity) {
-        v.reAlloc();
```

Comment:
Would you be able to help us, @lidavidm?

Do you see any reason why realloc isn't increasing value capacity there to get out of this while logic?

Reply:
Can you show a full self-contained example?

Reply:
sorry -- should have linked to the [issue](https://github.com/apache/arrow/issues/39810) which has sample code that shows the effect.

>I believe the Avro adapter fails on avro union types nested in records and arrays. The allocator tries to allocate increasingly more memory until it reaches the configured max (ie you can use -Darrow.vector.max_allocation_bytes=8589934592 to tell the allocator to use 1GB and it will try to allocate all of it for an empty array field with a nested union type field).
> `Memory required for vector capacity 508400 is (2097152), which is more than max allowed (1048576)`

Reply:
When run against `main`, the [test](https://github.com/darcysaum-toast/arrow/pull/3/files#diff-1672c2a7acda52586235fe3c8870dea23aaded7767756c066347c0294dac8a6eR478) in this PR demonstrates the issue.

Reply:
So it's something like `struct<union>`? I'd guess that reAlloc of a struct doesn't actually change its getValueCapacity.

https://github.com/apache/arrow/blob/90b30fcbfdfe12fa9ed497c3fa1cfe682b50168f/java/vector/src/main/java/org/apache/arrow/vector/complex/StructVector.java#L363-L371

So we take a min there, and super.getValueCapacity

https://github.com/apache/arrow/blob/90b30fcbfdfe12fa9ed497c3fa1cfe682b50168f/java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java#L344-L354

is based on the child's value capacity

Have you looked at this in a debugger? (I don't have the time to dig myself)

Reply:
> I'd guess that reAlloc of a struct doesn't actually change its getValueCapacity
that's right. 

yeah, i spent a bunch of time walking through this in the debugger and arrived at fact that the `min()` call you highlighted always returns 0 because `UnionVector::getValueCapacity` always returns 0 when the union is nested in either a record or an array.

[This](https://github.com/apache/arrow/blob/90b30fcbfdfe12fa9ed497c3fa1cfe682b50168f/java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java#L344-L354) is an interesting comparison. Are you suggesting that `getValueCapacity()` in `StructVector` should maybe look more like the implementation in `NonNullableStructVector`?

Reply:
Preferably UnionVector.getValueCapacity should be fixed